modules/selection/src/main/kotlin/com.r3.corda.lib.tokens.selection/memory/services/VaultWatcherService.kt

```diff
@@ -21,116 +21,194 @@ import net.corda.core.node.services.vault.QueryCriteria
+                        val owningKey = token.state.data.holder.owningKey
+                        lookupExternalIdFromKey(owningKey, appServiceHub)
                     }
+                }
             }
 
+            val existingStatesObservable = PublishSubject.create<Vault.Update<FungibleToken>>()
+
             val pageSize = 1000
             var currentPage = DEFAULT_PAGE_NUM
-            var (existingStates, observable) = appServiceHub.vaultService.trackBy(
+            val (firstPage, observable) = appServiceHub.vaultService.trackBy(
                     contractStateType = FungibleToken::class.java,
                     paging = PageSpecification(pageNumber = currentPage, pageSize = pageSize),
                     criteria = QueryCriteria.VaultQueryCriteria(),
                     sorting = sortByStateRefAscending())
-            val statesToProcess = mutableListOf<StateAndRef<FungibleToken>>()
-            while (existingStates.states.isNotEmpty()) {
-                statesToProcess.addAll(uncheckedCast(existingStates.states))
-                existingStates = appServiceHub.vaultService.queryBy(
-                        contractStateType = FungibleToken::class.java,
-                        paging = PageSpecification(pageNumber = ++currentPage, pageSize = pageSize)
-                )
+
+            val mergedObservable = existingStatesObservable.mergeWith(observable)
+            // we use the UPDATER thread for two reasons
+            // 1 this means we return the service before all states are loaded, and so do not hold up the node startup
+            // 2 because all updates to the cache (addition / removal) are also done via UPDATER, this means that until we have finished loading all updates are buffered preventing out of order updates
+            UPDATER.submit {
+                var shouldLoop = true
+                while (shouldLoop) {
+                    val newlyLoadedStates = appServiceHub.vaultService.queryBy(
+                            contractStateType = FungibleToken::class.java,
+                            paging = PageSpecification(pageNumber = ++currentPage, pageSize = pageSize),
+                            criteria = QueryCriteria.VaultQueryCriteria(),
+                            sorting = sortByStateRefAscending()
+                    )
+                    existingStatesObservable.onNext(Vault.Update(emptySet(), newlyLoadedStates.states.toSet()))
+                    shouldLoop = newlyLoadedStates.states.isNotEmpty()
+                }
             }
-            return TokenObserver(statesToProcess, uncheckedCast(observable), ownerProvider)
+            return TokenObserver(firstPage.states, uncheckedCast(mergedObservable), ownerProvider)
         }
     }
 
-    //owner -> tokenClass -> tokenIdentifier -> List
     init {
-        tokenObserver.initialValues.forEach(::addTokenToCache)
+        addTokensToCache(tokenObserver.initialValues)
         tokenObserver.source.subscribe(::onVaultUpdate)
```

Comment:
Maybe you can use RxJave `Schedulers.io()` here instead of the `newSingleThreadScheduledExecutor `?